Reduce memory usage by: only allocating arrays when needed, removing …

dgel opened the pull request:
…std::list usage, using short integer indices instead of pointers

The main algorithms used stays the same, including big-O complexity. These changes imply that individual identifiers are restricted to about 30k bytes, which I believe to be reasonable. In total, the 3 changes made here imply a 60-70% reduction in memory usage. I mentioned these changes in #474 , this PR does not include any usage of a more naive algorithm for identifying matches

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/475)
<!-- Reviewable:end -->

puremourning commented:
This sounds interesting ! However all the CI build/tests failed :(

dgel commented:
Yeah seems I've forgotten to add the changes to the tests -_-

vheon commented:
I'm also very much intrigued by this but I would love to see some benchmark written. This way we would be able to actually see by how much they differ in speed and actually what you've considered when optimizing this part of ycmd.

> These changes imply that individual identifiers are restricted to about 30k bytes,

Do you mean that each identifier must be under 30Kb? because actually identifier greater then 80 characters are not inserted in the index, see https://github.com/Valloric/ycmd/pull/370 for reference.

Valloric commented:
Like @vheon said, we already have a limit to max character length (80 characters). We're going to be improving our unicode support in the C++ layer so that's unlikely to be 80 _bytes_, but a multiple of that.

dgel commented:
Ok, so then these changes shouldn't impact any functionality =) I agree adding some performance benchmarking would be a good idea, but I'm not sure how best to add it to the project.
So far I've only moved some data around (behind pointers, removing pointers, and putting some data in structs instead of generic containers), and checked the runtime didn't deteriorate too badly in some synthetic tests. I was using some c++11 to time it though, and it didn't collect any detailed statistics..
Is this project using any testing framework?

Also, I'm not sure why the appveyor build fails. I'm not very familiar with CMake, and am assuming it's generating the VS project files that are failing atm, but have no idea how to fix it. Can someone lend me a hand with this?

micbou commented:
> Also, I'm not sure why the appveyor build fails. I'm not very familiar with CMake, and am assuming it's generating the VS project files that are failing atm, but have no idea how to fix it. Can someone lend me a hand with this?

You need to export the `LetterNodeArray::~LetterNodeArray` and `LetterNode &LetterNodeArray::operator[]` symbols for the tests on Windows by using the `YCM_DLL_EXPORT` macro.

vheon commented:
I've started to look at the code. The algorithm is not the same, am I correct?

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cpp/ycm/Candidate.cpp, line 100 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KGnPU6TxfFfN8EUI6nu:-KGnPU6TxfFfN8EUI6nv:-434676362) ([raw file](https://github.com/valloric/ycmd/blob/b0fc8460b00af30bbdd47415e67e1afeb2034e47/cpp/ycm/Candidate.cpp#L100)):*
you can move this line before `if ( case_sensitive && IsUppercase( letter ) ) {` and remove the `else node = NULL` here and in the `else` branch.

---

*[cpp/ycm/Candidate.h, line 62 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KGnPhSmP5XKlzqqOmN5:-KGnPhSmP5XKlzqqOmN6:1864489241) ([raw file](https://github.com/valloric/ycmd/blob/b0fc8460b00af30bbdd47415e67e1afeb2034e47/cpp/ycm/Candidate.h#L62)):*
This doesn't seem to be defined anywhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KGnPMlR_wR6Re_JPyh-:747815014)*
<!-- Sent from Reviewable.io -->

dgel commented:
The algorithm should be the same in general. I've changed some small things though: the original was saving pointers to subsequent occurrences of a character in an array of lists, but then only used the first occurrence and the first occurrence of the uppercase version in code that was using it. I changed it by only storing indices instead of pointers, and only storing the first occurrence of the character and first occurrence of the uppercase version of the character in a dedicated struct.

Other than that, I only moved some things behind pointers or the other way around. The matching algorithm should work the same way 

---

Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[cpp/ycm/Candidate.cpp, line 100 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KGnPU6TxfFfN8EUI6nu:-KGs2k-A8DHZggCw4qd_:1758359711) ([raw file](https://github.com/valloric/ycmd/blob/b0fc8460b00af30bbdd47415e67e1afeb2034e47/cpp/ycm/Candidate.cpp#L100)):*
Fixed.

---

*[cpp/ycm/Candidate.h, line 62 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KGnPhSmP5XKlzqqOmN5:-KGs2Xd7vwZaqeQ5PPG0:2005257455) ([raw file](https://github.com/valloric/ycmd/blob/b0fc8460b00af30bbdd47415e67e1afeb2034e47/cpp/ycm/Candidate.h#L62)):*
Yeah that wasn't supposed to be in this PR. Removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KGs2o9JyXqGlm6lFTCq:965642507)*
<!-- Sent from Reviewable.io -->

vheon commented:
> The algorithm should be the same in general. I've changed some small things though: the original was saving pointers to subsequent occurrences of a character in an array of lists, but then only used the first occurrence and the first occurrence of the uppercase version in code that was using it. I changed it by only storing indices instead of pointers, and only storing the first occurrence of the character and first occurrence of the uppercase version of the character in a dedicated struct.

Thanks for explanation :)

>  but then only used the first occurrence and the first occurrence of the uppercase version in code that was using it

So if this is true (I admit I don't remember how it works by heart :sob:), could we simplify the original structure by using only two pointers instead of all the pointers if we decide to not change the algorithm? I have to refresh this part of ycmd carefully to help here :(

Anyway first we should concentrate on benchmark `ycmd` as it is to get a baseline

> I was using some c++11 to time it though, and it didn't collect any detailed statistics..
Is this project using any testing framework?

We use Google Test for the C++ part. What were you using? I'm sure we are open to other frameworks for testing if they are helpful.

---

Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KGsEBev0EAwp-ty6-xP:-586478224)*
<!-- Sent from Reviewable.io -->

dgel commented:
> So if this is true (I admit I don't remember how it works by heart :sob:), could we simplify the original structure by using only two pointers instead of all the pointers if we decide to not change the algorithm? I have to refresh this part of ycmd carefully to help here :(

Yeah the original used a list, but I've changed it to a struct of two short integers, storing offsets into the LetterNode array instead of pointers. So, that change would have been made in this PR already

> We use Google Test for the C++ part. What were you using? I'm sure we are open to other frameworks for testing if they are helpful.

I'm sorry, I meant a benchmarking framework or library. Something that figures out how many times to run a benchmark to get meaningful statistics, can give mean and std dev etc. I'm not very knowledgeable in these things personally, I tend to just run some things in loops and compare the processing time. For optimizing the memory in this PR, I just had ycm use a very large tags file instead of doing any formal tests. 

How would you want to proceed with this? add some benchmarking in a separate ticket and then revisit this PR? I'm not sure how long it would take me to add a library / framework to the CMake stuff, it seemed pretty complex

puremourning commented:
Google have a pretty cool benchmarking API we could use

vheon commented:
> Google have a pretty cool benchmarking API we could use
 
@puremourning you mean https://github.com/google/benchmark?? I've never used it personally but I've seen talks about using it (but I don't think that count 😝 )

Valloric commented:
I've personally used the google/benchmark code (before it was open sourced) several times when I was at Google. It's simple and awesome. We should use it, though the benchmarks should probably be a separate PR.

dgel commented:
I've finally taken the time to do some simple benchmarks. I find the results very encouraging.

I've decided to just do the benchmarks locally, using a c++11 framework. For testing the memory usage I just used htop and a program that creates a large amount of LetterNodes.
If anyone is willing, please run these tests on your machine as well, results will vary across compilers I think..

please find the result and code snippets here:

https://gist.github.com/dgel/0ee4709015288ead82c14e2c079710bd

Valloric commented:
If those benchmarks are sound (and I don't see a reason why they wouldn't be), this is some awesome shit! Twice as fast to match while using six times less memory? Yes, please!

Valloric commented:
Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[cpp/ycm/Candidate.cpp, line 102 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjsMfb7FnBnzypUuj5:-KIjsMfb7FnBnzypUuj6:-1958641702) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/Candidate.cpp#L102)):*
> ```cpp
>     } else {
>       if ( nearest->eitherIndex >= 0 )
>         node = ( *root_node_ )[nearest->eitherIndex];
> ```

Spaces inside brackets needed.

---

*[cpp/ycm/LetterNode.h, line 50 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjsYk-2cm1a4bqnGPG:-KIjsYk-2cm1a4bqnGPH:-1190960218) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNode.h#L50)):*
> ```objc
>   }
> 
>   inline void AddNodeForLetter( char letter, short index ) {
> ```

I'd rather not have this marked inline; put it in the cpp file and let the compiler inline it if necessary. I'm more willing to let inline accessors slide, but this does not fall into that category.

---

*[cpp/ycm/LetterNode.h, line 64 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjsnqW0DVvbZI16_ky:-KIjsnqXj4QfKLP1dz3Z:-1275421533) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNode.h#L64)):*
> ```objc
>   }
> 
>   inline LetterNode *operator[]( short index ) {
> ```

Use an int instead of a short. The short isn't going to be any faster; ints are the register-native size.

---

*[cpp/ycm/LetterNodeArray.cpp, line 1 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjtVGMjN7dW6QhWEAy:-KIjtVGMjN7dW6QhWEAz:849869257) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L1)):*
> ```cpp
> // Copyright (C) 2011, 2012 Google Inc.
> ```

This copyright header is wrong now. Make it "2016 ycmd contributors".

---

*[cpp/ycm/LetterNodeArray.cpp, line 29 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjtbdQ35kC7fnVmo_v:-KIjtbdQ35kC7fnVmo_w:-1947720351) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L29)):*
> ```cpp
>   : size_( text.size() ) {
>   // properly aligned type to hold elements
>   typedef boost::aligned_storage<sizeof( LetterNode ), boost::alignment_of<LetterNode>::value>::type storage_type;
> ```

This needs to be 80 columns.

---

*[cpp/ycm/LetterNodeArray.cpp, line 29 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjztBlGUakG3bx6HNZ:-KIjztBlGUakG3bx6HN_:287858454) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L29)):*
> ```cpp
>   : size_( text.size() ) {
>   // properly aligned type to hold elements
>   typedef boost::aligned_storage<sizeof( LetterNode ), boost::alignment_of<LetterNode>::value>::type storage_type;
> ```

What's the benefit of going with this complex custom storage vs a plan std::vector that has been correctly sized in the call to its ctor? The vector has a pointer (as this does), an int-sized size (as this does) and an int-sized capacity. If the only thing we save is the capacity, then I'd rather go with vector since its simpler to maintain. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

dgel commented:
Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[cpp/ycm/LetterNode.h, line 50 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjsYk-2cm1a4bqnGPG:-KIxAA8hwSfDYz01tvCS:1801997100) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNode.h#L50)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> I'd rather not have this marked inline; put it in the cpp file and let the compiler inline it if necessary. I'm more willing to let inline accessors slide, but this does not fall into that category.
</details>
The compiler won't inline anything that's compiled into a .o file -- inline (as an instructrion to inline something) is just a suggestion to the compiler that they usually ignore. It's marked inline only because the other functions here were marked so, but I can remove the inline keyword (they do effectively nothing here), and possibly move this function to the .cpp. It won't ever be inlined if I do so however 

---

*[cpp/ycm/LetterNode.h, line 64 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjsnqW0DVvbZI16_ky:-KIxAg-be7izRzcAdcrb:2141440492) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNode.h#L64)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Use an int instead of a short. The short isn't going to be any faster; ints are the register-native size.
</details>
Sure. 

---

*[cpp/ycm/LetterNodeArray.cpp, line 1 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjtVGMjN7dW6QhWEAy:-KIxBoEkrwSCh6u_Gjnx:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L1)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> This copyright header is wrong now. Make it "2016 ycmd contributors".
</details>
Done.

---

*[cpp/ycm/LetterNodeArray.cpp, line 29 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjztBlGUakG3bx6HNZ:-KIxBs-XBjQiQ-YVC70Z:1733270618) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L29)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> What's the benefit of going with this complex custom storage vs a plan std::vector that has been correctly sized in the call to its ctor? The vector has a pointer (as this does), an int-sized size (as this does) and an int-sized capacity. If the only thing we save is the capacity, then I'd rather go with vector since its simpler to maintain. 
</details>
the vector would have a size_t size and capacity actually, not int, the two differ on some platforms
That said, I didn't actually implement it for the capacity, I did it because LetterNodes are noncopyable, so
can't be used with std::vector at the moment. Could simply make LetterNodes copyable of course, I just 
wasn't sure if there was an important reason they were made noncopyable.

If there aren't any pressing reasons to have LetterNodes be noncopyable, I'll just remove this class

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[cpp/ycm/LetterNodeArray.cpp, line 29 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjztBlGUakG3bx6HNZ:-KIxMnUACPMI_VkSlbNr:-1568497907) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L29)):*
<details><summary><i>Previously, dgel wrote…</i></summary>
> the vector would have a size_t size and capacity actually, not int, the two differ on some platforms
> That said, I didn't actually implement it for the capacity, I did it because LetterNodes are noncopyable, so
> can't be used with std::vector at the moment. Could simply make LetterNodes copyable of course, I just 
> wasn't sure if there was an important reason they were made noncopyable.
> 
> If there aren't any pressing reasons to have LetterNodes be noncopyable, I'll just remove this class
> 
</details>
I would prefer to remove the `noncopyable` and use `std::vector`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[cpp/ycm/LetterNode.h, line 50 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjsYk-2cm1a4bqnGPG:-KJ3bclyPpkTsSjUi822:252702097) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNode.h#L50)):*
<details><summary><i>Previously, dgel wrote…</i></summary>
> The compiler won't inline anything that's compiled into a .o file -- inline (as an instructrion to inline something) is just a suggestion to the compiler that they usually ignore. It's marked inline only because the other functions here were marked so, but I can remove the inline keyword (they do effectively nothing here), and possibly move this function to the .cpp. It won't ever be inlined if I do so however 
</details>
I hear your concern about it not being inlined, but it would indeed be inlined in LTO builds. I don't think we do LTO builds right now, but we could. In general, I tend to strongly lean towards going with the simpler approach first and avoiding premature optimization. 

So let's put it in the cpp file for now.

---

*[cpp/ycm/LetterNodeArray.cpp, line 1 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjtVGMjN7dW6QhWEAy:-KJ3lI9lKlTD-8EJclQH:222839791) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L1)):*
<details><summary><i>Previously, dgel wrote…</i></summary>
> Done.
</details>
Maybe you didn't push the change? I don't see it.

---

*[cpp/ycm/LetterNodeArray.cpp, line 29 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjztBlGUakG3bx6HNZ:-KJ3lSFf3otxbhs4W9Uk:-1008378665) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L29)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> I would prefer to remove the `noncopyable` and use `std::vector`
</details>
Yeah I don't think there was a specific reason for them to be non-copyable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

dgel commented:
Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[cpp/ycm/LetterNodeArray.cpp, line 1 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjtVGMjN7dW6QhWEAy:-KJH2PB0ju7VJ6tB4Hnb:2123938020) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L1)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Maybe you didn't push the change? I don't see it.
</details>
Removed this file, no need for license now =)

---

*[cpp/ycm/LetterNodeArray.cpp, line 29 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KIjztBlGUakG3bx6HNZ:-KJH2U0jlgWORA_84Gxb:560107473) ([raw file](https://github.com/valloric/ycmd/blob/151a78478ba60be08705a080a06dd86d253e4510/cpp/ycm/LetterNodeArray.cpp#L29)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Yeah I don't think there was a specific reason for them to be non-copyable.
</details>
Fair enough, removed this class

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

vheon commented:
I gave this a little read and left some comments. I'm still try to fully understand the new structure/algorithm so I'm sorry If I left some stupid comments :stuck_out_tongue_winking_eye: 

<details><summary><i>Previously, coveralls wrote…</i></summary>
> 
> [![Coverage Status](https://coveralls.io/builds/6378799/badge)](https://coveralls.io/builds/6378799)
> 
> Coverage increased (+0.2%) to 84.957% when pulling **10d5f50d6021a6d284c421486a7ffa02d392dcb2 on dgel:memory-usage** into **f3dd8fd70a04343914a97b2ebabb4ac877c4ea12 on Valloric:master**.
> 
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[cpp/ycm/Candidate.cpp, line 73 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJH9_-irnL6vjWPk_b1:-KJH9_-j6dzdlJTDkkWj:-986823278) ([raw file](https://github.com/valloric/ycmd/blob/f3dd8fd70a04343914a97b2ebabb4ac877c4ea12/cpp/ycm/Candidate.cpp#L73)):*
> ```cpp
> }
> ```

we actually use 2 spaces between member functions :smiley: 

---

*[cpp/ycm/Candidate.cpp, line 96 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJH9grPc3qwSUaH4VLk:-KJH9grQbqutQ0yUUPLc:1422984132) ([raw file](https://github.com/valloric/ycmd/blob/f3dd8fd70a04343914a97b2ebabb4ac877c4ea12/cpp/ycm/Candidate.cpp#L96)):*
> ```cpp
> }
> ```

same as above

---

*[cpp/ycm/LetterNode.cpp, line 34 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHCdVIqMqYHz8e8D2k:-KJHCdVIqMqYHz8e8D2l:-1135747727) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNode.cpp#L34)):*
> ```cpp
>     is_uppercase_( false ) {
> 
>   letternode_per_text_index_.reserve(text.size());
> ```

spaces inside the parenthesis

---

*[cpp/ycm/LetterNode.cpp, line 38 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHFnlNDAWUHFySymxG:-KJHFnlOaZucdJ7gBgm3:-1127805909) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNode.cpp#L38)):*
> ```cpp
>   for ( uint i = 0; i < text.size(); ++i ) {
>     letternode_per_text_index_.push_back( LetterNode( text[ i ], i ) );
>     AddNodeForLetter( text[ i ], i );
> ```

From what I understood now we store only the indexes of the nearest LetterNode for each letter but here we're calling `AddNodeForLetter` on each character of the `text`. Wouldn't we override the index of the nearest `LetterNode` with furthest one here? 

---

*[cpp/ycm/LetterNode.cpp, line 50 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHBtLqieI9V82Jzag7:-KJHBtLqieI9V82Jzag8:-1676359682) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNode.cpp#L50)):*
> ```cpp
> void LetterNode::AddNodeForLetter( char letter, short index ) {
>   if ( IsUppercase( letter ) ) {
>     if ( letters_[ letter ].upperIndex == -1 )
> ```

couldn't we keep a reference of the result of `letters_[ letter ]` and use it through the function instead of calling the `operator[]` four times?

---

*[cpp/ycm/LetterNodeListMap.cpp, line 50 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHB0K9iAIj9uf7RhIG:-KJHB0KAQw2m0GZt7ZyK:1029231068) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNodeListMap.cpp#L50)):*
> ```cpp
> }
> 
> LetterNodeListMap::~LetterNodeListMap() {
> ```

I might be missing something here, but why keep the empty destructor?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KJHGi_Gf5PZdEiO3P9D:983067981)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[cpp/ycm/Candidate.cpp, line 73 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJH9_-irnL6vjWPk_b1:-KJIXmcMLTZrkzfhxEk2:568351987) ([raw file](https://github.com/valloric/ycmd/blob/f3dd8fd70a04343914a97b2ebabb4ac877c4ea12/cpp/ycm/Candidate.cpp#L73)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> we actually use 2 spaces between member functions :smiley: 
</details>
@vheon meant two empty lines. Please don't leave two space characters. :D

---

*[cpp/ycm/LetterNodeListMap.h, line 43 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJIYFzNtei3zsaw_7QR:-KJIYFzNtei3zsaw_7QS:-638935161) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNodeListMap.h#L43)):*
> ```objc
>   {}
> 
>   short eitherIndex;
> ```

These fields need more docs explaining what they're for. Possibly more verbose names too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

dgel commented:
Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[cpp/ycm/Candidate.cpp, line 73 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJH9_-irnL6vjWPk_b1:-KKcetX4X0o0u0vvpHMM:-140808748) ([raw file](https://github.com/valloric/ycmd/blob/f3dd8fd70a04343914a97b2ebabb4ac877c4ea12/cpp/ycm/Candidate.cpp#L73)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> @vheon meant two empty lines. Please don't leave two space characters. :D
</details>
Haha ok, using 2 newlines now =)

---

*[cpp/ycm/Candidate.cpp, line 96 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJH9grPc3qwSUaH4VLk:-KKcezrxQmpQR0H3cve4:677588487) ([raw file](https://github.com/valloric/ycmd/blob/f3dd8fd70a04343914a97b2ebabb4ac877c4ea12/cpp/ycm/Candidate.cpp#L96)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> same as above
</details>
done

---

*[cpp/ycm/LetterNode.cpp, line 34 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHCdVIqMqYHz8e8D2k:-KKcf0V0Q_KQn__rEygl:-467915377) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNode.cpp#L34)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> spaces inside the parenthesis
</details>
fixed

---

*[cpp/ycm/LetterNode.cpp, line 38 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHFnlNDAWUHFySymxG:-KKcf3mdL62uoVxwX4X6:1506045952) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNode.cpp#L38)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> From what I understood now we store only the indexes of the nearest LetterNode for each letter but here we're calling `AddNodeForLetter` on each character of the `text`. Wouldn't we override the index of the nearest `LetterNode` with furthest one here? 
</details>
Hmm AddNodeForLetter is probably poorly named, it only adds if a letter hasn't been added yet. Will rename to something more explicit

---

*[cpp/ycm/LetterNode.cpp, line 50 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHBtLqieI9V82Jzag7:-KKcgXqm_OchTN3HGubd:-1790877027) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNode.cpp#L50)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> couldn't we keep a reference of the result of `letters_[ letter ]` and use it through the function instead of calling the `operator[]` four times?
</details>
yes thanks, done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[cpp/ycm/LetterNode.cpp, line 38 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHFnlNDAWUHFySymxG:-KKcj8tNdeRQCqEbKhGz:-451858737) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNode.cpp#L38)):*
<details><summary><i>Previously, dgel wrote…</i></summary>
> Hmm AddNodeForLetter is probably poorly named, it only adds if a letter hasn't been added yet. Will rename to something more explicit
</details>
Now it makes more sense. Thanks ;)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

dgel commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[cpp/ycm/LetterNodeListMap.cpp, line 50 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJHB0K9iAIj9uf7RhIG:-KKcl3kiJJF9m_GZ54gQ:-1518740514) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNodeListMap.cpp#L50)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> I might be missing something here, but why keep the empty destructor?
</details>
Uhm don't know =P removed it

---

*[cpp/ycm/LetterNodeListMap.h, line 43 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KJIYFzNtei3zsaw_7QR:-KKckKdf8Z82ft5Tqxyo:526112712) ([raw file](https://github.com/valloric/ycmd/blob/b787a6e79b62e1aa8af7a8040ef66e19a15c686a/cpp/ycm/LetterNodeListMap.h#L43)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> These fields need more docs explaining what they're for. Possibly more verbose names too.
</details>
Added explanation and renamed fields

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

Valloric commented:
I think this is good now! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Can't wait for this to land. :)

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KKdL47U6c2tkHY1bXPE:692383722)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Took a really brief look over the code. I'm not super familiar with this part of the codebase so nothing insightful.

There is of course the issue that the build is failing in CI...

---

Reviewed 1 of 1 files at r2, 2 of 4 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cpp/ycm/Candidate.cpp, line 103 \[r8\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KKdMEhAM9fE-7eSdzlG:-KKdMEhAM9fE-7eSdzlH:1549980559) ([raw file](https://github.com/valloric/ycmd/blob/fa78e89030972bd2fda20622be36cf54cf99a7c7/cpp/ycm/Candidate.cpp#L103)):*
> ```cpp
>     if ( case_sensitive && IsUppercase( letter ) ) {
>       if ( nearest->indexOfFirstUppercaseOccurrence >= 0 )
>         node =  ( *root_node_ )[ nearest->indexOfFirstUppercaseOccurrence ];
> ```

there is  1 extra whitespace character after `=`

---

*[cpp/ycm/tests/LetterNode_test.cpp, line 36 \[r8\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KKdNVPuhPlayvMMRsqA:-KKdNVPuhPlayvMMRsqB:1715363485) ([raw file](https://github.com/valloric/ycmd/blob/fa78e89030972bd2fda20622be36cf54cf99a7c7/cpp/ycm/tests/LetterNode_test.cpp#L36)):*
> ```cpp
>                       Property( &LetterNode::LetterIsUppercase, false ) ) );
> 
>   const NearestLetterNodeIndices *nearest_nodes = root_node.NearestLetterNodesForLetter( 'i' );
> ```

does this sneak over 80 chars? hard to tell in reviewable interface

---

*[cpp/ycm/tests/LetterNode_test.cpp, line 37 \[r8\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KKdNbkcM1AjiLwy60iX:-KKdNbkcM1AjiLwy60iY:1349009131) ([raw file](https://github.com/valloric/ycmd/blob/fa78e89030972bd2fda20622be36cf54cf99a7c7/cpp/ycm/tests/LetterNode_test.cpp#L37)):*
> ```cpp
> 
>   const NearestLetterNodeIndices *nearest_nodes = root_node.NearestLetterNodesForLetter( 'i' );
>   EXPECT_THAT( root_node[nearest_nodes->eitherIndex],
> ```

space inside `[` and `]` (a few instances of that in this file)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KKdO9JgDQUBgaPg-is3:863106007)*
<!-- Sent from Reviewable.io -->

vheon commented:
@dgel you forgot to rename the fields in the tests :)

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KKdRPBLomJMMBASdNic:-696159394)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 1 of 1 files at r2, 1 of 6 files at r5, 2 of 4 files at r7, 2 of 3 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

dgel commented:
@vheon thanks, I think it's fixed now

---

Review status: 4 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[cpp/ycm/Candidate.cpp, line 103 \[r8\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KKdMEhAM9fE-7eSdzlG:-KKdwuCNtw8F9qDVkETj:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/fa78e89030972bd2fda20622be36cf54cf99a7c7/cpp/ycm/Candidate.cpp#L103)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary>
> there is  1 extra whitespace character after `=`
</details>
Done.

---

*[cpp/ycm/tests/LetterNode_test.cpp, line 36 \[r8\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KKdNVPuhPlayvMMRsqA:-KKdwvqPTJTXRvSzg8sL:-1319937293) ([raw file](https://github.com/valloric/ycmd/blob/fa78e89030972bd2fda20622be36cf54cf99a7c7/cpp/ycm/tests/LetterNode_test.cpp#L36)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary>
> does this sneak over 80 chars? hard to tell in reviewable interface
</details>
yeah it did, fixed it now

---

*[cpp/ycm/tests/LetterNode_test.cpp, line 37 \[r8\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KKdNbkcM1AjiLwy60iX:-KKdxA2kSeMveCVueW9n:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/fa78e89030972bd2fda20622be36cf54cf99a7c7/cpp/ycm/tests/LetterNode_test.cpp#L37)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary>
> space inside `[` and `]` (a few instances of that in this file)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KKdxH-6HM9QYFJ_ZkQf:856510636)*
<!-- Sent from Reviewable.io -->

micbou commented:
I tested [your benchmarks](https://gist.github.com/dgel/0ee4709015288ead82c14e2c079710bd#ycmd-benchmarks) and I am obtaining similar results so this looks excellent.

Since this PR greatly reduces memory usage and also improves performance, I think we need to bump the library version. @dgel Could you increment the version number in the `CORE_VERSION` file?

---

Reviewed 1 of 6 files at r5, 2 of 4 files at r7, 2 of 3 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[cpp/ycm/LetterNode.cpp, line 25 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KLCLxHTPhOrrxQcAJMh:-KLCLxHUxCBYHLBRLRcX:-1577068405) ([raw file](https://github.com/valloric/ycmd/blob/38c59b3eb0bf35fde79fd66bc0f28da69488f587/cpp/ycm/LetterNode.cpp#L25)):*
> ```cpp
> 
> LetterNode::LetterNode( char letter, int index )
>   : index_( index ) ,
> ```

Extra space before comma.

---

*[cpp/ycm/LetterNode.h, line 58 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KLCMQs-HoDKCswqdW6V:-KLCMQs-HoDKCswqdW6W:-871883860) ([raw file](https://github.com/valloric/ycmd/blob/38c59b3eb0bf35fde79fd66bc0f28da69488f587/cpp/ycm/LetterNode.h#L58)):*
> ```objc
> 
>   inline LetterNode *operator[]( int index ) {
>     return &letternode_per_text_index_[index];
> ```

Missing spaces between square brackets.

---

*[cpp/ycm/LetterNodeListMap.cpp, line 55 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KLCMheFAJVbzjes0Dl8:-KLCMheGCnAgRwPbbuHP:1971661962) ([raw file](https://github.com/valloric/ycmd/blob/38c59b3eb0bf35fde79fd66bc0f28da69488f587/cpp/ycm/LetterNodeListMap.cpp#L55)):*
> ```cpp
>   if ( !letters_ ) {
>     letters_.reset( new NearestLetterNodeArray() );
>   }
> ```

We can remove the curly brackets here.

---

*[cpp/ycm/LetterNodeListMap.h, line 75 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KLCNWj2G66mZAKRCSb3:-KLCNWj3JLGPpCFxcoH8:-498485081) ([raw file](https://github.com/valloric/ycmd/blob/38c59b3eb0bf35fde79fd66bc0f28da69488f587/cpp/ycm/LetterNodeListMap.h#L75)):*
> ```objc
> 
> private:
>   typedef boost::array<NearestLetterNodeIndices , NUM_LETTERS> 
> ```

Trailing space.

---

*[cpp/ycm/tests/LetterNode_test.cpp, line 36 \[r8\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KKdNVPuhPlayvMMRsqA:-KLCNa5OdlyUmTN1EeI0:2029921564) ([raw file](https://github.com/valloric/ycmd/blob/fa78e89030972bd2fda20622be36cf54cf99a7c7/cpp/ycm/tests/LetterNode_test.cpp#L36)):*
<details><summary><i>Previously, dgel wrote…</i></summary>
> yeah it did, fixed it now
</details>
Still above 80 characters.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KLCO2uZPwCGIg_-E-zH:477154850)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with @micbou's comments. Thanks again!

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KLCRwqOVwTRvNKkDMXO:-598375308)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 1 of 6 files at r5, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Agreed on the core version bump. This PR is going to make lots of people very happy!

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KLDVkUEMlGq5dnNfW72:-1649139314)*
<!-- Sent from Reviewable.io -->

dgel commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[cpp/ycm/LetterNode.cpp, line 25 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KLCLxHTPhOrrxQcAJMh:-KLkYg3CX2qQaAP6ZY0T:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/38c59b3eb0bf35fde79fd66bc0f28da69488f587/cpp/ycm/LetterNode.cpp#L25)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Extra space before comma.
</details>
Done.

---

*[cpp/ycm/LetterNode.h, line 58 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KLCMQs-HoDKCswqdW6V:-KLkYeMxJ8A7STqgUZ2A:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/38c59b3eb0bf35fde79fd66bc0f28da69488f587/cpp/ycm/LetterNode.h#L58)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Missing spaces between square brackets.
</details>
Done.

---

*[cpp/ycm/LetterNodeListMap.cpp, line 55 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KLCMheFAJVbzjes0Dl8:-KLkZ-fINkaGsLop-aeA:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/38c59b3eb0bf35fde79fd66bc0f28da69488f587/cpp/ycm/LetterNodeListMap.cpp#L55)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> We can remove the curly brackets here.
</details>
Done.

---

*[cpp/ycm/LetterNodeListMap.h, line 75 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KLCNWj2G66mZAKRCSb3:-KLkZ4b4xMkqfc2wXUqK:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/38c59b3eb0bf35fde79fd66bc0f28da69488f587/cpp/ycm/LetterNodeListMap.h#L75)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Trailing space.
</details>
Done.

---

*[cpp/ycm/tests/LetterNode_test.cpp, line 36 \[r8\]](https://reviewable.io:443/reviews/valloric/ycmd/475#-KKdNVPuhPlayvMMRsqA:-KLkZLfmCqgbXSa1Mmrw:903995768) ([raw file](https://github.com/valloric/ycmd/blob/fa78e89030972bd2fda20622be36cf54cf99a7c7/cpp/ycm/tests/LetterNode_test.cpp#L36)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Still above 80 characters.
</details>
woops, *actually* fixed now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@micbou @vheon This is a big change, so while this PR has two LGTM's, could we get a third here? Feel free to trigger homu when satisfied.

---

Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KLl_irTQkGrnmeEWmOJ:-1993321094)*
<!-- Sent from Reviewable.io -->

vheon commented:
I already did the review on the majority of this some time ago :+1: 

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> @homu r=Valloric

---

Reviewed 1 of 2 files at r9, 6 of 6 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/475#-:-KLlkxbmPLa6KZZdVxei:1362150872)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 4d9d0f1 has been approved by `Valloric`

<!-- @homu r=Valloric 4d9d0f183a7208ecfbf0c1f838c1251b65981da9 -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/ycmd/builds/141987657)
